### PR TITLE
Support FMA operation

### PR DESCRIPTION
**Pull request by thvnx**
This commit adds the floating-point fused multiply-add operation (FMA) to the Pervasives module.

The following changes are made:
- _configure_: check `math.h` for the C99 `fma()` operation.
- fma declarations in _pervasives.ml[i]_ (stdlib and otherlibs/threads).
- C `fma()` call or emulation in _byterun/floats.c_.
- Dedicated tests.

I am a regular ocaml follower, as well as an intensive FMA user. I've done this work to enjoy both ocaml language and FMA operation. If useful to the community, could this contribution be integrated to the next ocaml release?

The test suite passes successfully.

### CODE REVIEW

**Comment by thvnx** [comment]
When emulated (only activated if C99 functions aren't provided by math.h), the FMA doesn't resist to overflows or underflows. Do I have to mention it in documentation?

**Comment by alainfrisch** [comment]
Can you confirm that this is intended to improve the precision of the calculation, not to make the combined operation faster (which wouldn't probably be the case because of the C function call)?

**Comment by thvnx** [comment]
You are right. It's for accuracy purposes, and additionally, improves the programming of extended-precision-based numerical algorithms (with perhaps performance improvements in some such cases).

It's very usefull for me but maybe not so relevant for the ocaml developers community?

**Comment by alainfrisch** [comment]
I've a biais towards numerical code, so I'll let other core maintainers comment on the relevance of including the function in the stdlib.  (The fact that it could at some point be implemented directly in the backend to benefit from CPU instructions would be an argument in favor of including the function.)

Anyway, if it is decided that the function deserves to be in the stdlib, it should probably go in the future Float module discussed elsewhere, not in Pervasives.

**Comment by thvnx** [comment]
Thanks for your comments. I agree.

Where can I find the discussion about the Float module?

**Comment by xavierleroy** [comment]
I agree with @alainfrisch that this would be a fine addition to the future `Float` module.  (Perhaps along with other C99 FP functions that we still miss.)

The fallback implementation is taken from authors I trust.  However it uses some C99 features (declaration in the middle of a block) that (some version of) MSVC doesn't support.

It should be easy to have `ocamlopt` turn calls to this function into FMA instructions provided by the processor, if they exist.  E.g. for PowerPC and for ARM64, possibly for ARM too.  For x86-64 FMA instructions were added to SSE only recently, so we have a portability issue.

**Comment by alainfrisch** [comment]
Adding a Float module was discussed in #964, and this is now conditioned on #1010.

**Comment by jhjourdan** [comment]
Is the fallback implementation correct if overflow occur? The paper which is cited explicitly exclude this case from its proofs.

I am not sure whether this applies here, but note that overflow could occur in intermediate computations but not in the correct final result. In that case, it is even not enough to add something like "undefined result in case of overflow" in the documentation, because then it is not clear what overflow mean.

We should also take care about infinites and NaN...

**Comment by thvnx** [comment]
@jhjourdan, the fallback implementation isn't correct if overflow occurs. In that case, the result will be as if computations were done without FMA.

I'll look if a more robust FMA emulation exists (maybe [here](https://sourceware.org/git/?p=glibc.git;a=blob_plain;f=sysdeps/ieee754/dbl-64/s_fma.c;hb=HEAD)).

EDIT: emulation linked above seems conform to IEEE standard.

**Comment by thvnx** [comment]
Does anyone understand why AppVeyor build fails? I don't have any Windows environment to reproduce the actual issue, i.e. warnings from line 1768 to 1773.

**Comment by nojb** [comment]
This is related to some recent PR merges and it is being worked on.  It should be resolved in the near future.

**Comment by xavierleroy** [comment]
Concerning overflows, ISO C99 7.12.13.1 says "A range error may occur", without further details.  So, I'm not sure all `fma` implementations out there handle overflow in `x * y` correctly.

**Comment by thvnx** [comment]
About the FMA emulation: the one provided by the GLIBC is conform to IEEE 754 requirements (since the version 2.24, see [here](https://sourceware.org/bugzilla/show_bug.cgi?id=13304)). Note that it uses the same algorithm which is currently implemented in this PR. I will soon adapt the fallback consequently.

**Comment by jhjourdan** [comment]
Great !

**Comment by thvnx** [comment]
I've updated the FMA emulation and added new tests. FMA fallback supports now overflow and almost all underflows (some double-rounding issues are very difficult to handle: 3 of 317 tests still fail).

**Comment by damiendoligez** [comment]
Now we have merged #1010 and #1638, so we have a `Float` module. @thvnx could you update this PR?

**Comment by thvnx** [comment]
Great! I'll do it ASAP.

**Comment by thvnx** [comment]
I moved the fma operation in the Float module.

But, I still have a few things to do:

- [x] 1. Make tests working again. testsuite/HACKING.adoc is incomplete, directives for 'Creating a new test' are missing. I do not understand how to add new tests well for now.
- [ ] 2. In case of fma emulation only (when C99 functions are not available), some double-rounding issues are very difficult to handle: 3 of 317 tests still fail.
- [ ] 3. For the tests, I've taken all the numerical values from the glibc ones, do I need to be more specific in testsuite/tests/fma/fma.ml (with a link to the original file for example)?

**Comment by damiendoligez** [comment]
> `testsuite/HACKING.adoc` is incomplete, directives for 'Creating a new test' are missing. I do not understand how to add new tests well for now.

cc @shindere

**Comment by shindere** [comment]
Well I have to say I never paid attention to `testsuite/HACKING.adoc`',
sorry about that. I just looked at it and it seems it's a work in
progress.

@thvnx did you try to look at existing tests to see how they work?

If you are able to describe the kind of test you would like to add, I
can describe how you should do.

**Comment by thvnx** [comment]
@shindere, yes, I have a .ml file with its .reference counterpart. I've added an ocamltests file which contains the file name of my .ml test. When I run 'make tests', I obtain the following output:

```
List of directories returning no results:
    tests/fma
```

**Comment by shindere** [comment]
You need to add an ocamltest block at the beginning of your file. For
instance

```ocaml
(* TEST *)
```

may be enough. You can run that test directly by doing, from the
directory in question:

```bash
../../../ocamltest/ocamltest test.ml
```

You can also run the testsuite just on this directory by running the
following command from the testsuite directory:

```bash
make one DIR=tests/fma
```

**Comment by thvnx** [comment]
Thanks, It's done!

**Comment by thvnx** [comment]
It's almost done, I still have to fix the FMA emulation fallback to pass all the remaining failing tests.

**Comment by thvnx** [comment]
I did not succeed in solving the three failed tests for the hard-to-tackle double-rounding issue (in the case of fallback emulation). So, I just disabled them. It works well when FMA hardware is provided but not fully IEEE compliant in emulation mode. OK anyway?

File: stdlib/float.mli
Comment (alainfrisch) [inline stdlib/float.mli]: I'm not sure "deprecated" is relevant here.  Something like "If no hardware support is provided, ...".  But can this also be caused by a lack of libc support?  And if the libc provides the emulation itself without hardware support, is it fully accurate?

**Comment by alainfrisch** [comment]
>  OK anyway?

Ok for me, but perhaps @xavierleroy or @Chris00 would like to comment on it.

File: stdlib/float.mli
Comment (thvnx) [reply stdlib/float.mli]: It depends on the libc. Glibc is IEEE compliant, I don't know about the others. The fallback I wrote is very close to the one of the glibc (but they can switch rounding modes and other things that I can't do here).

File: stdlib/float.mli
Comment (alainfrisch) [inline stdlib/float.mli]: Ok, so IIUC, the criterion is not so much that the hardware support is provided or not, but whether the libc provides a compliant version (emulated or hardware based) or not.  There can be cases where hardware support is available but emulation is still used (your emulation, or from the libc).  (Sorry to nitpick, but it's important to explain as precisely possible sources of differences in fp calculations across different machines.)

Comment (thvnx) [reply stdlib/float.mli]: Yes, you're absolutely right! I'll fix it in the next commit.

File: stdlib/float.mli
Comment (alainfrisch) [inline stdlib/float.mli]: Proposal:
````
[fma x y z] returns [x * y + z], with a best effort for computing this expression with a single rounding, using either hardware instructions (providing full IEEE compliance) or a software emulation.

@since 4.08
````

Comment (thvnx) [reply stdlib/float.mli]: Fixed, thanks!

File: configure
Comment (alainfrisch) [inline configure]: You should include `fma` in the list of functions tested with hasgot2.

File: stdlib/float.mli
Comment (alainfrisch) [inline stdlib/float.mli]: 1. 4.08
2. I did not think about it before, but what about saying a word about performance (with and without emulation) compared to the naive version?  People might think that using fma is a good idea for performance reason even when precision is not a big concern, and it's not obvious whether this is a good idea.  Is it true that with hardware support, fma would be fastest but with emulation from libc or yours, itwould be slower?

File: Changes
Comment (alainfrisch) [inline Changes]: Please include people involved in the review.

File: testsuite/tests/fma/fma.ml
Comment (alainfrisch) [inline testsuite/tests/fma/fma.ml]: I'll let you decide, but perhaps, for those tests, you could specify either a list of possible results (the correct one, and the one obtained by the emulation), or do a check "modulo epsilon"; just to make sure that the function is not too crazy in those cases.

**Review by alainfrisch — APPROVED** [review]
Minor comments only.

I'm not competent to check the emulation code but @xavierleroy said he had enough trust, so I think this can be merged once comments are addressed unless someone else speaks up.

File: Changes
Comment (thvnx) [reply Changes]: Reviewers only or all participants?

Comment (alainfrisch) [reply Changes]: Looking quickly at the discussion, I'd say : @jhjourdan @xavierleroy @alainfrisch

File: testsuite/tests/fma/fma.ml
Comment (thvnx) [reply testsuite/tests/fma/fma.ml]: Fixed (with list of possible results), integration in progress, I'm waiting for mscv64 failures if any.

File: configure
Comment (thvnx) [reply configure]: Done

File: stdlib/float.mli
Comment (thvnx) [reply stdlib/float.mli]: Add a note about performance.

File: Changes
Comment (thvnx) [reply Changes]: Thanks, done!

File: Changes
Comment (alainfrisch) [inline Changes]:
```suggestion
  (Laurent Thévenoux, review by Alain Frisch, Jacques-Henri Jourdan,
```

Comment (thvnx) [reply Changes]: Sorry!

**Comment by alainfrisch** [comment]
Thanks @thvnx for your contribution and patience when dealing with incremental comments!

**Comment by shindere** [comment]
This PR causes a failure of the fma test on Inria's CI (Cygwin32 machine).
The problem happens wiht and without flambda in native and bytecode mode.
There is a difference between the expected output and the obtained one:

```
@@ -252,17 +252,17 @@
 252 OK!
 253 OK!
 254 OK!
-255 OK!
+255 FAIL!	fma (0x1.fffp+0, 0x1.0000000000001p+0, -0x1.fffp+0) returned 0x1p-51 instead of 0x1.fffp-52.
 256 OK!
 257 OK!
-258 OK!
-259 OK!
-260 OK!
+258 FAIL!	fma (0x1.deadbeef2feedp+1023, 0x0.deadbeef2feedp-1022, -0x1.a05f8c01a4bfbp+1) returned 0x1.098p-53 instead of 0x1.0989687bc9da4p-53.
+259 FAIL!	fma (0x1.deadbeef2feedp+900, 0x0.deadbeef2feedp-1022, -0x1.a05f8c01a4bfbp-122) returned 0x1.098p-176 instead of 0x1.0989687bc9da4p-176.
+260 FAIL!	fma (0x1.fffffffffffffp+1023, 0x1.001p+0, -0x1.fffffffffffffp+1023) returned 0x1p+1012 instead of 0x1.fffffffffffffp+1011.
 261 OK!
 262 OK!
 263 OK!
-264 OK!
-265 OK!
+264 FAIL!	fma (0x1.deadbeef2feedp-495, 0x1.deadbeef2feedp-495, -0x1.bf86a5786a574p-989) returned 0x0.00000428p-1022 instead of 0x0.0000042625a1fp-1022.
+265 FAIL!	fma (0x1.deadbeef2feedp-503, 0x1.deadbeef2feedp-503, -0x1.bf86a5786a574p-1005) returned 0x0.000000000428p-1022 instead of 0x0.0000000004262p-1022.
 266 OK!
 267 OK!
 268 OK!
@@ -295,10 +295,10 @@
 295 OK!
 296 OK!
 297 OK!
-298 OK!
-299 OK!
-300 OK!
-301 OK!
+298 FAIL!	fma (0x1.fffffffffffffp-1, 0x1.fffffffffffffp-1, -0x1.ffffffffffffep-1) returned 0x0p+0 instead of 0x1p-106.
+299 FAIL!	fma (0x1.fffffffffffffp-1, -0x1.fffffffffffffp-1, 0x1.ffffffffffffep-1) returned 0x0p+0 instead of -0x1p-106.
+300 FAIL!	fma (-0x1.fffffffffffffp-1, 0x1.fffffffffffffp-1, 0x1.ffffffffffffep-1) returned 0x0p+0 instead of -0x1p-106.
+301 FAIL!	fma (-0x1.fffffffffffffp-1, -0x1.fffffffffffffp-1, -0x1.ffffffffffffep-1) returned 0x0p+0 instead of 0x1p-106.
 302 OK!
 303 OK!
 304 OK!
```

**Comment by thvnx** [comment]
@shindere , I can easily fix it. Do I have to create a new pull request or I can just pull in this one?

**Comment by alainfrisch** [comment]
No unfortunately, you need to create a new PR.  Is the problem a bad software emulation from that libc?

**Comment by thvnx** [comment]
@alainfrisch, OK. Yes, I guess this is due to another software support for the fma on this platform (or something related to subnormals).

@shindere, I'll create a new PR with this patch: https://github.com/thvnx/ocaml/commit/06d120cd98d2c62955d09734cdb655ed0ce5b51b. Can you test it before someone merge it to trunk?

See #2120

**Comment by shindere** [comment]
For several weeks, Inria's CI was broken in several ways, so that
the testsuite could not be run for the Cygwin 64 port.

Now that this has been fixed, it turns out that the FMA tests fail on
this platform, for both native and bytecode, with and without flambda.

More precisely, in all these cases the test's output differs from the
expected one:

```
@@ -255,15 +255,15 @@
 255 OK!
 256 OK!
 257 OK!
-258 OK!
-259 OK!
-260 OK!
-261 OK!
-262 OK!
+258 FAIL!	fma (0x1.deadbeef2feedp+1023, 0x0.deadbeef2feedp-1022, -0x1.a05f8c01a4bfbp+1) returned 0x0p+0 instead of 0x1.0989687bc9da4p-53.
+259 FAIL!	fma (0x1.deadbeef2feedp+900, 0x0.deadbeef2feedp-1022, -0x1.a05f8c01a4bfbp-122) returned 0x0p+0 instead of 0x1.0989687bc9da4p-176.
+260 FAIL!	fma (0x1.fffffffffffffp+1023, 0x1.001p+0, -0x1.fffffffffffffp+1023) returned infinity instead of 0x1.fffffffffffffp+1011.
+261 FAIL!	fma (-0x1.fffffffffffffp+1023, 0x1.fffffffffffffp+0, 0x1.fffffffffffffp+1023) returned -infinity instead of -0x1.ffffffffffffdp+1023.
+262 FAIL!	fma (0x1.fffffffffffffp+1023, 0x1p+1, -0x1.fffffffffffffp+1023) returned infinity instead of 0x1.fffffffffffffp+1023.
 263 OK!
-264 OK!
-265 OK!
-266 OK!
+264 FAIL!	fma (0x1.deadbeef2feedp-495, 0x1.deadbeef2feedp-495, -0x1.bf86a5786a574p-989) returned 0x0p+0 instead of 0x0.0000042625a1fp-1022.
+265 FAIL!	fma (0x1.deadbeef2feedp-503, 0x1.deadbeef2feedp-503, -0x1.bf86a5786a574p-1005) returned 0x0p+0 instead of 0x0.0000000004262p-1022.
+266 FAIL!	fma (0x1p-537, 0x1p-538, 0x0.0000000000001p-1022) returned 0x0.0000000000001p-1022 instead of 0x0.0000000000002p-1022.
 267 OK!
 268 OK!
 269 OK!
@@ -275,8 +275,8 @@
 275 OK!
 276 OK!
 277 OK!
-278 OK!
-279 OK!
+278 FAIL!	fma (0x0.0000000000001p-1022, 0x1p-1, 0x0.fffffffffffffp-1022) returned 0x0.fffffffffffffp-1022 instead of 0x1p-1022.
+279 FAIL!	fma (-0x0.0000000000001p-1022, 0x1p-1, -0x0.fffffffffffffp-1022) returned -0x0.fffffffffffffp-1022 instead of -0x1p-1022.
 280 OK!
 281 OK!
 282 OK!

```

And configure says:

```
checking for fma... yes
```

Any idea what may go wrong here?

**Comment by dra27** [comment]
@shindere - see #2177

**Comment by shindere** [comment]
David Allsopp (2019/03/01 00:52 -0800):
> @shindere - see #2177

Yes sorry, indeed! Actually I think that one was waiting for you to take
a comment I did into account or just react to it saying you won't take
it into account ;-)

**Pull request merged**
